rfcs/20211207-softmax-v2/README.md

```diff
@@ -0,0 +1,192 @@
+softmax primitive.
+
+Another long-lasting request is so-called `masked` softmax. It comes from
+Transformer-based models when Decoder applies Encoder-Decoder Attention layer.
+At this point the machine is not allowed to see future tokens of the sentence,
+that's why this "mask" should hide the future starting from the processed token.
+Mask makes softmax input data shorter.
+
+Due to the nice property of softmax that data on the left end of range does not
+contribute much or at all (depending on the range) to final answer and likely
+vanishes, there is a trick to perform this optimization on the user side, which
+will be explained shortly. Thus, usually matmul precedes by softmax. The mask
+can be incorporated into bias which would fill specific spots with `-inf` or
+`INT32_MIN` values and fuse its addition into matmul. Then softmax will handle a
+full, non-masked, length but those points will not contribute to the final
+answer and will not spoil the accuracy of the model. Such trick allows to handle
+a desired behavior much faster than softmax could do it.
+
+Solution via softmax primitive has API, integration, and implementation
+complexity. That's why masked feature won't be added to softmax primitive as
+well.
+
+## Proposal
+
+### Option 1
+
+There are no options to cover different memory descriptors and algorithms other
+than creating a new operation descriptor, `dnnl_softmax_v2_desc_t`. It will be
+extended over previous one with `dnnl_alg_kind_t alg_kind`,
+`dnnl_memory_desc_t dst_desc` and `dnnl_memory_desc_t diff_dst_desc`.
+`data_desc` analogues will be replaced with `src_desc` correspondently.
+Implementation details will be similar to pooling_v2.
+
+#### C API
+
+```c
+/// dnnl_types.h
+
+/// Kinds of primitives. Used to implement a way to extend the library with new
+/// primitives without changing the ABI.
+typedef enum {
+    ...
+    /// A softmax version 2 primitive (softmax with destination memory
+    /// descriptor and algorithm kind).
+    dnnl_softmax_v2,
+    ...
+} dnnl_primitive_kind_t;
+
+/// Kinds of algorithms.
+typedef enum {
+    ...
+    /// Softmax
+    dnnl_softmax_alg,
+    /// Logsoftmax
+    dnnl_logsoftmax_alg,
+} dnnl_alg_kind_t;
+
+/// A descriptor of a Softmax operation.
+typedef struct {
+    /// The kind of primitive. Used for self-identifying the primitive
+    /// descriptor. Must be #dnnl_softmax.
+    dnnl_primitive_kind_t primitive_kind;
+    /// The kind of propagation. Possible values: #dnnl_forward_training and
+    /// #dnnl_forward_inference.
+    dnnl_prop_kind_t prop_kind;
+    /// Source and destination memory descriptor.
+    dnnl_memory_desc_t src_desc;
+    /// Source and Destination of gradient memory descriptor.
+    dnnl_memory_desc_t diff_src_desc;
+    /// The axis along which to perform the softmax.
+    int softmax_axis;
+    /// Softmax algorithm. Possible values: #dnnl_softmax and
+    /// #dnnl_logsoftmax.
+    dnnl_alg_kind_t alg_kind;
```

Comment:
It seems that we put `alg_kind` after `prop_kind` in most operation descriptors.

Reply:
It's true, but I don't want to mess with casting since internally we would treat pointers of different types and I need members be at exact same spots where they are. Same thing as it was with pooling.

Reply:
I'm fine with the slightly different order but just to clarify. I'd expect that inside the library we always use `dnnl_softmax_v2_desc_t`. 
I can only think of the primitive cache where you might want to implement a hash/comparison function for `dnnl_softmax_desc_t` that would compare all old fields and for `dnnl_softmax_v2_desc_t` that would compare the new fields.

Reply:
Yes, we do operate with v2 inside only. The problem happens when user still uses v1 API, queries for v1 desc and we have to return the original desc. We achieve it with using desc() (which is v2) but keeping fields compatible (the way I left them). This is the primary reason for such order IIRC. Maybe there's something else...

Reply:
Yeah, I forgot about that. Thanks for clarifying it.